Index: Color3.cs
===================================================================
--- Color3.cs	(revision 47)
+++ Color3.cs	(working copy)
@@ -754,7 +754,7 @@
         /// </returns>
         public override string ToString()
         {
-            return string.Format(CultureInfo.CurrentCulture, "Red:{1} Green:{2} Blue:{3}", Red, Green, Blue);
+            return string.Format(CultureInfo.CurrentCulture, "Red:{0} Green:{1} Blue:{2}", Red, Green, Blue);
         }
 
         /// <summary>
@@ -769,7 +769,7 @@
             if (format == null)
                 return ToString();
 
-            return string.Format(CultureInfo.CurrentCulture, "Red:{1} Green:{2} Blue:{3}", Red.ToString(format, CultureInfo.CurrentCulture),
+            return string.Format(CultureInfo.CurrentCulture, "Red:{0} Green:{1} Blue:{2}", Red.ToString(format, CultureInfo.CurrentCulture),
                 Green.ToString(format, CultureInfo.CurrentCulture), Blue.ToString(format, CultureInfo.CurrentCulture));
         }
 
@@ -782,7 +782,7 @@
         /// </returns>
         public string ToString(IFormatProvider formatProvider)
         {
-            return string.Format(formatProvider, "Red:{1} Green:{2} Blue:{3}", Red, Green, Blue);
+            return string.Format(formatProvider, "Red:{0} Green:{1} Blue:{2}", Red, Green, Blue);
         }
 
         /// <summary>
Index: Design/FieldPropertyDescriptor.cs
===================================================================
--- Design/FieldPropertyDescriptor.cs	(revision 47)
+++ Design/FieldPropertyDescriptor.cs	(working copy)
@@ -46,7 +46,7 @@
         }
 
         public FieldPropertyDescriptor(FieldInfo fieldInfo)
-            : base(fieldInfo.Name, (Attribute[])fieldInfo.GetCustomAttributes(true))
+            : base(fieldInfo.Name, Array.ConvertAll(fieldInfo.GetCustomAttributes(true), item=> item as Attribute))
         {
             this.fieldInfo = fieldInfo;
         }
Index: Design/Vector3Converter.cs
===================================================================
--- Design/Vector3Converter.cs	(revision 47)
+++ Design/Vector3Converter.cs	(working copy)
@@ -100,8 +100,15 @@
         /// </exception>
         public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
         {
-            var values = ConvertToValues<float>(context, culture, value);
-            return values != null ? new Vector3(values) : base.ConvertFrom(context, culture, value);
+            try
+            {
+                var values = ConvertToValues<float>(context, culture, value);
+                return values != null ? new Vector3(values) : base.ConvertFrom(context, culture, value);
+            }
+            catch(Exception e)
+            {
+                throw e;
+            }
         }
 
         /// <summary>
Index: Half.cs
===================================================================
--- Half.cs	(revision 47)
+++ Half.cs	(working copy)
@@ -122,6 +122,7 @@
         public ushort RawValue
         {
             get { return value; }
+            set { this.value = value; }
         }
 
         /// <summary>
